tests/administrators/test_data.py

```diff
@@ -0,0 +1,169 @@
+        ids=[
+            "force_reset",
+            "groups",
+        ],
+    )
+    async def test(
+        self,
+        all_permissions,
+        mongo,
+        no_permissions,
+        snapshot,
+        static_time,
+        update,
+        data_layer,
+        fake2,
+    ):
+        """
+        Ensure updates are applied correctly.
+
+        """
+
+        user_1 = await fake2.users.create()
+
+        await gather(
+            mongo.groups.insert_many(
+                [
+                    {
+                        "_id": "peasants",
+                        "permissions": {**no_permissions, "create_sample": True},
+                    },
+                    {
+                        "_id": "kings",
+                        "permissions": {**no_permissions, "create_ref": True},
+                    },
+                ],
+                session=None,
+            ),
+            mongo.users.update_one(
+                {"_id": user_1.id}, {"$set": {"groups": ["peasants"]}}
+            ),
+        )
+
+        assert await data_layer.administrators.update_user(
+            user_1.id, update
+        ) == snapshot(name="obj")
+        assert await mongo.users.find_one() == snapshot(
+            name="db", exclude=props("password")
+        )
+
+    async def test_password(self, fake2, mongo, snapshot, data_layer, static_time):
+        """
+        Test editing an existing user.
+
+        """
+        user_1 = await fake2.users.create()
+
+        assert await data_layer.administrators.update_user(
+            user_1.id, UpdateUserRequest(password="hello_world")
+        ) == snapshot(name="obj")
+
+        document = await mongo.users.find_one()
+
+        assert document == snapshot(name="db", exclude=props("password"))
+
+        # Ensure the newly set password validates.
+        assert await validate_credentials(mongo, user_1.id, "hello_world")
+
+    async def test_does_not_exist(self, data_layer):
+        with pytest.raises(ResourceNotFoundError) as err:
+            await data_layer.administrators.update_user(
+                "user_id", UpdateUserRequest(administrator=False)
+            )
+            assert "User does not exist" == str(err)
+
+
+@pytest.mark.parametrize("role", [None, AdministratorRole.BASE, AdministratorRole.FULL])
+async def test_update_role(fake2, mongo, snapshot, data_layer, role, static_time):
+    """
+    Test changing the administrator role of a user.
+
+    """
+    user_1 = await fake2.users.create()
+
+    assert await data_layer.administrators.update_role(
+        user_1.id, UpdateAdministratorRoleRequest(role=role)
+    ) == snapshot(name="obj")
+
+    updated_user = await mongo.users.find_one()
+    assert updated_user["administrator"] == (role == AdministratorRole.FULL)
+
+    expected_list = [(user_1.id, role)] if role is not None else []
+
+    assert (
+        await data_layer.administrators._authorization_client.list_administrators()
+        == expected_list
+    )
```

Comment:
I can switch it snapshots, but I find the DX of diffing more than one per test suboptimal. The expected result of this test was sufficiently simple that to me it made more sense to predict the expected result rather than use snapshots. Main upside I can see of snapshots is that it would flag a change in what the role is, which could be good if that change is borked, otherwise it is annoying.

Happy to change it though if you want snapshots use throughout